header/p2p/exchange.go

```diff
@@ -130,22 +139,34 @@ func (ex *Exchange) performRequest(
 	if err != nil {
 		return nil, err
 	}
+	if err = stream.SetWriteDeadline(time.Now().Add(writeDeadline)); err != nil {
+		log.Warnf("error setting deadline: %s", err)
+	}
 	// send request
 	_, err = serde.Write(stream, req)
 	if err != nil {
 		stream.Reset() //nolint:errcheck
 		return nil, err
 	}
+	err = stream.CloseWrite()
+	if err != nil {
+		log.Warn(err)
+	}
 	// read responses
 	headers := make([]*header.ExtendedHeader, req.Amount)
 	for i := 0; i < int(req.Amount); i++ {
 		resp := new(header_pb.ExtendedHeader)
+		if err = stream.SetReadDeadline(time.Now().Add(readDeadline)); err != nil {
+			log.Warnf("error setting deadline: %s", err)
+		}
 		_, err := serde.Read(stream, resp)
 		if err != nil {
 			stream.Reset() //nolint:errcheck
 			return nil, err
 		}
-
+		if err = stream.SetReadDeadline(time.Time{}); err != nil {
```

Comment:
Do we need this?

Reply:
Shouldn't `SetReadDeadline` override the previous read deadline?

Reply:
I think no. Did not find such mechanism: 
here we set a deadline https://github.com/libp2p/go-yamux/blob/781eed82a74eb7b3868c390fce8091c3c40c708d/deadline.go#L31 , that will close done channel at #L55 when timeout is reached. So, I assumed that we will get here https://github.com/libp2p/go-yamux/blob/781eed82a74eb7b3868c390fce8091c3c40c708d/stream.go#L115 every time once deadline will be reached

Reply:
Ack, I see that's how libp2p does it too.

Reply:
you can resolve this @vgonkivs

Reply:
Sorry for writing in closed thread, but I had the exact same question. I've digged into libp2p and found it is quite interesting piece of code. There are several places where this situation is handled very smart.

1. If we set new deadline before previous one was reached [#L55](https://github.com/libp2p/go-yamux/blob/781eed82a74eb7b3868c390fce8091c3c40c708d/deadline.go#L55) could be handled by line [#L35](https://github.com/libp2p/go-yamux/blob/781eed82a74eb7b3868c390fce8091c3c40c708d/deadline.go#L35). `!d.timer.Stop()` call will stop `time.AfterFunc` timer and will prevent it from executing callback. So new deadline will override previous one.

2. Another case is if deadline has happen after `Read()` was called , meaning `time.AfterFunc` has fired and closed the `d.closed` channel. If we try to set new deadline by `stream.SetReadDeadline()` the `d.closed` needs to be recreated. Which will happen in line  [#L52](https://github.com/libp2p/go-yamux/blob/781eed82a74eb7b3868c390fce8091c3c40c708d/deadline.go#L52) 

3. The more interesting part is the fact that potentially `time.AfterFunc` could fired, but didn't finished closing `d.closed` channel yet (imagine there is `time.Sleep` inside callback before `close(d.close)`) . In this case `d.timer.Stop()` wouldn't stop internal timer because it is already stoped and will return `false`. So we wait `d.close` to be closed  inside `time.AfterFunc` callback in [#L36](https://github.com/libp2p/go-yamux/blob/781eed82a74eb7b3868c390fce8091c3c40c708d/deadline.go#L36). Then proceed to [#L52](https://github.com/libp2p/go-yamux/blob/781eed82a74eb7b3868c390fce8091c3c40c708d/deadline.go#L52)  path.

So seems like the line could be removed but won't hurt having it either.